doc/designs/quic-design/quic-connID-retire.md

```diff
@@ -0,0 +1,71 @@
+QUIC Route Requirements
+=======================
+
+* Two connection ID -- one local, one remote
+
+Zero Length Connection ID
+-------------------------
+
+If a remote uses a zero length connection ID, it is not allowed to retire it
+or to create new IDs.  Produce an error if such occurs.
+
+Each zero length ID must be associated with a different local address and
+port.
+
+Retiring Connection ID
+----------------------
+
+When a remote asks to retire a connection ID (RETIRE_CONNECTION_ID) we have to:
+
+* Flag the route(s) as retired
+* ACK the packet (I think this happens here & doesn't finalise the retirement)
+* Wait for all outstanding packets on the route(s) to be sent & acked
+  * On a retransmit, we could migrate the packets to a still active connection
+    but we don't need too
+  * We could rewrite some (but probably not all) packets to use a different
+    route.
+  * Paragraph 5 in 5.1.2 implies rewriting is required.
+* Should respond with a NEW_CONNECTION_ID frame retiring the route(s)
+  * Not sure if it is mandatory to send a retirement.
+  * RETIRE_CONNECTION_ID is probably an alternative but this reduces the
+    number of concurrent connections so not ideal.
+* Delete the connections
+
+The standard says, in multiple places, that initiating a retirement is done
+with the RETIRE_CONNECTION_ID frame.  Paragraph 4 in 5.1.2 implies that it can
+also be done with a NEW_CONNECTION_ID frame.  Suggest we accept either.
```

Comment:
I think this is a confusion. The connection IDs of a peer are issued by NEW_CONNECTION_ID frame and that frame also contains the information which previously issued IDs should be retired by the peer within the Retire Prior To field of the frame. The RETIRE_CONNECTION_ID frame is then sent by the other peer when it is not going to use the retired connection ID anymore.

In MVP we use zero length connection id as the local connection ID -> we do not issue new connection ids and we do not expect to receive RETIRE_CONNECTION_ID frames as we do not issue any other IDs other than the zero length one - i.e., received RETIRE_CONNECTION_ID frame can be either ignored or treated as PROTOCOL_VIOLATION connection error (the RFC allows both).

On the other hand in MVP we need to be able to handle NEW_CONNECTION_ID frames with newly issued connection ids of our peer. To avoid the need for full fledged connection id cache in MVP we could send active_connection_id_limit transport parameter set to 1. This means the peer can issue only one new connection ID to us at the same time retiring the prior one. We then need to send RETIRE_CONNECTION_ID as an acknowledgement that we switched to use the new id and drop using the previous one.

Reply:
This is discussing _retirement_ not new connections.

I think active_connection_id_limit should be set to 1.  It's permissible to create a new connection ID beyond this limit so long as the included retirements bring it back under.  It's also suggested that up to twice this number of retired connections should be supported.  Even though we're received a retirement from our peer, we can't retire that connection until we flush everything using it -- this means we need to keep the retired IDs around for a period of time.

Reply:
Actually RFC requires the active_connection_id_limit to be at least 2.

Reply:
> Even though we're received a retirement from our peer, we can't retire that connection until we flush everything using it -- this means we need to keep the retired IDs around for a period of time.

That's not true - there is nothing that "uses the connection ID" anymore - we are talking here about the destination connection id and that is used only when transmitting packets. We can always start using the new connection ID immediately. However we cannot retire the old one unless peer asks for it, otherwise we would risk creating a feedback loop of NEW_CONNECTION_ID/RETIRE_CONNECTION_ID frame ping-pong.

Reply:
I'm working on implementation.

Reply:
Any queued packets need to either be sent with the old ID (which is completely legitimate) or rewritten with the new ID which is more difficult but possible.  It's the demarkation betwixt the two that is problematic.

Reply:
If we queue a RETIRE_CONNECTION_ID frame using the CFQ, this obviously can't be sent until everything else already in our TXE queue is drained to the network. So the only real issue is datagram reordering in the network. Even if reordering does occur the worst that can happen is the peer ignores incoming datagrams and they have to get retransmitted.

I've just checked another QUIC implementation and it appears it queues a RETIRE_CONNECTION_ID to a CFQ-like construct immediately on processing a NEW_CONNECTION_ID frame. So this seems to be a viable implementation strategy in practice.

Reply:
Implementation of what I propose in #20892

Reply:
Lost packets will also cause "old' ID to be sent won't they?  Or do we retransmit with the new ID?

Reply:
The lost packets are always re-created so they'll use new ID.

Reply:
Yeah. A fine distinction to be made with QUIC is that QUIC never actually retransmits packets, only frames (potentially with more up-to-date information).